enterprise/cmd/frontend/internal/executorqueue/handler/multihandler.go

```diff
@@ -0,0 +1,211 @@
+	"strings"
+
+	"github.com/sourcegraph/log"
+	"golang.org/x/exp/slices"
+
+	btypes "github.com/sourcegraph/sourcegraph/enterprise/internal/batches/types"
+	uploadsshared "github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/uploads/shared"
+	executorstore "github.com/sourcegraph/sourcegraph/enterprise/internal/executor/store"
+	executortypes "github.com/sourcegraph/sourcegraph/enterprise/internal/executor/types"
+	"github.com/sourcegraph/sourcegraph/internal/workerutil"
+	dbworkerstore "github.com/sourcegraph/sourcegraph/internal/workerutil/dbworker/store"
+	"github.com/sourcegraph/sourcegraph/lib/api"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type MultiHandler struct {
+	JobTokenStore         executorstore.JobTokenStore
+	CodeIntelQueueHandler QueueHandler[uploadsshared.Index]
+	BatchesQueueHandler   QueueHandler[*btypes.BatchSpecWorkspaceExecutionJob]
+	logger                log.Logger
+}
+
+func NewMultiHandler(
+	jobTokenStore executorstore.JobTokenStore,
+	codeIntelQueueHandler QueueHandler[uploadsshared.Index],
+	batchesQueueHandler QueueHandler[*btypes.BatchSpecWorkspaceExecutionJob],
+) MultiHandler {
+	return MultiHandler{
+		JobTokenStore:         jobTokenStore,
+		CodeIntelQueueHandler: codeIntelQueueHandler,
+		BatchesQueueHandler:   batchesQueueHandler,
+		logger:                log.Scoped("executor-multi-queue-handler", "The route handler for all executor queues"),
+	}
+}
+
+var validQueues = []string{"batches", "codeintel"}
+
+func validateQueues(queues []string) []string {
+	var invalidQueues []string
+	for _, queue := range queues {
+		if !slices.Contains(validQueues, queue) {
+			invalidQueues = append(invalidQueues, queue)
+		}
+	}
+	return invalidQueues
+}
+
+// TODO: fairly sure this is basically executortypes.DequeueRequest with Queues extended
+type dequeueRequest struct {
+	Queues       []string `json:"queues"`
+	ExecutorName string   `json:"executorName"`
+	Version      string   `json:"version"`
+	NumCPUs      int      `json:"numCPUs,omitempty"`
+	Memory       string   `json:"memory,omitempty"`
+	DiskSpace    string   `json:"diskSpace,omitempty"`
+}
+
+func (m *MultiHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var req dequeueRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		err = errors.Wrap(err, fmt.Sprintf("Failed to unmarshal payload"))
+		m.logger.Error(err.Error())
+		m.marshalAndRespondError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	if err := validateWorkerHostname(req.ExecutorName); err != nil {
+		m.logger.Error(err.Error())
+		m.marshalAndRespondError(w, err, http.StatusBadRequest)
+	}
+
+	version2Supported := false
+	if req.Version != "" {
+		var err error
+		version2Supported, err = api.CheckSourcegraphVersion(req.Version, "4.3.0-0", "2022-11-24")
+		if err != nil {
+			m.marshalAndRespondError(w, err, http.StatusBadRequest)
+			return
+		}
+	}
+
+	if invalidQueues := validateQueues(req.Queues); len(invalidQueues) != 0 {
```

Comment:
Personal nit, I prefer `>` over `!=` with slice lengths

```suggestion
	if invalidQueues := validateQueues(req.Queues); len(invalidQueues) > 0 {
```